File: link.pl
#!/usr/bin/env perl
#use File::Copy qw(copy);
use Cwd 'abs_path';
use Getopt::Std;
use File::Basename;
use File::Copy;

my $opt_string='f';
getopts("$opt_string", \%opt);

$force=1 if $opt{f};

$homedir=$ENV{HOME};
$sourcedir=dirname(abs_path($0));

$manifest="$sourcedir/manifest.conf";

open (FH,$manifest) or die "Cannot read manifest file manifest.conf!\n";
$forced=true if $opt{f};

print "* Re-creating files \n";
print "** FORCE OVERWRITE **\n" if $forced;
print "* Target = $homedir \n";
print "* Source = $sourcedir \n\n";

while (<FH>) {
	chomp;
	next if /^\#/;
	next if /^\s*$/;

	my ($action, $source, $target)=split /\s+/;

	$target =~ s/~/$homedir/;
	$source = "$sourcedir/$source";

	warn "Cannot read $source" if ! -r $source;

	if ($action eq "symlink") {
		printf "- Linking $source to $target\n";
		if ( -l "$target" ) { 
			unlink $target or warn "X Existing $target link cannot be removed.\n"; 
		} elsif ( -e "$target" && $forced) { 
			local $backup = "$target".".orig";
			print "O File Exists -- Renaming $target to $backup.\n";
			move($target,$backup) or warn "X Cannot rename $target to $backup.\n";
		} elsif ( -e "$target" ) {
			warn "X Existing $target is a file and will not be removed."; 
			next;
		}

		symlink($source, $target) or warn "X Cannot symlink $source to $target\n";

	} elsif ($action eq "copy") {
		printf "- Copying $source to $target\n";
		if ( -l "$target" ) {
			unlink $target or warn "X Existing $target link cannot be removed.\n"; 
		} elsif ( -e "$target" && $forced) {
			print "O File Exists -- Renaming $target to $backup.\n";
			move($target,$backup) or warn "X Cannot rename $target to $backup.\n";
		} elsif ( -e "$target" ) {
			warn "X Existing $target is a file and will not be removed."; 
			next;
		}

		copy($source, $target) or warn "X Cannot copy $source to $target \n";	
	} else {
		warn "X No clue on how to $action $source to $target \n";
	}	
}

File: link.py
#!/usr/bin/env python3
import os
import shutil
import sys

# Parse command line options
force = '-f' in sys.argv

# Set up directories and files
homedir = os.path.expanduser('~')
sourcedir = os.path.dirname(os.path.abspath(__file__))
manifest = os.path.join(sourcedir, 'manifest.conf')

# Check if manifest file exists
if not os.path.isfile(manifest):
    print("Cannot read manifest file manifest.conf!")
    sys.exit(1)

print("* Re-creating files")
if force:
    print("** FORCE OVERWRITE **")
print("* Target =", homedir)
print("* Source =", sourcedir, "\n")

# Process each line in the manifest file
with open(manifest) as f:
    for line in f:
        line = line.strip()
        if line.startswith('#') or not line:
            continue

        action, source, target = line.split()
        target = target.replace('~', homedir)
        source = os.path.join(sourcedir, source)

        if not os.access(source, os.R_OK):
            print(f"Cannot read {source}")
            continue

        print(f"- {action.capitalize()}ing {source} to {target}")

        if os.path.exists(target):
            if force:
                backup = target + ".orig"
                print(f"O File Exists -- Renaming {target} to {backup}.")
                shutil.move(target, backup)
            else:
                print(f"X Existing {target} is a file and will not be removed.")
                continue

        if action == "symlink":
            os.symlink(source, target)
        elif action == "copy":
            shutil.copy(source, target)
        else:
            print(f"X No clue on how to {action} {source} to {target}")
